### ISSUE TITLE
Notifications - Implementing custom group notifications

### ISSUE THREAD

**Issue by Johann-Snapp** [RocketChat/Rocket.Chat#2192]
As said in #1315, @geekgonecrazy doesn't like when people notify @all, and he's not the only one.
Plus, it's comfortable when you can call @frontends, @backends, @lawyers, ect.

This issue is not about asking a feature, it's more about how to implement it correctly.
So, let's go !

- In order to configure it, we'll need to persist custom mentions.

- As admins could configure it, what happens in case of conflict between a username & a mention-group name ?

- We'll need to update the mentions autocomplete popup

- What happens when we click on the mentionned group ? Does it open the right panel with user list ? How does it work for @all ? Does it shows the channel's user list or does it send data to the panel ?

- In which admin section do we put it ? Users ? Channels ? Message ? New section ?

- Maybe we can implement the confirmation popup when notifying many people: #909

- Maybe we could add aliases to mentions so that we search by name or alias.

Did I forget something ? Is there someone motivated to help me working on it ? Is there someone already working on it ?

**Comment by damienmollard** [comment]
+1

**Comment by Johann-Snapp** [comment]
Found this: https://github.com/RocketChat/Rocket.Chat/pull/1564

**Comment by engelgabriel** [comment]
This will need to tied up with the ROLES, so you could mention `@rolename`

**Comment by Johann-Snapp** [comment]
If we do this, we have to setup a exclude list in order to not have things like @bot :D

**Comment by engelgabriel** [comment]
@channel will send a notification to all team members in the channel where you post the message.

@everyone notifies every member of your team. This command can only be used in the team-wide channel (usually called #general).

Both notifications will be sent to all team members in the channel whether they are active, away or offline.

@here is best for sending announcements relevant only to team members that are _currently working or available_.

**Comment by marceloschmidt** [comment]
I'll add a reference to #2135 as this is an important guide if we are going on that direction.

**Comment by nemolivier** [comment]
On this I would like : 
* @chans : notify all chans (only for admin, when reboot, update, etc.)
* @admins / @modos : add an attention sign in front of the chan to notify a problem in chan (someone trolling, need for moderation, etc.)

**Comment by engelgabriel** [comment]
See
https://github.com/RocketChat/Rocket.Chat/issues/1315
https://github.com/RocketChat/Rocket.Chat/issues/2135

**Comment by entwu** [comment]
+1
It would be awesome if I could call role/group member ! e.g. @testers

**Comment by jgtoriginal** [comment]
Actually I'd go for dynamic tags, for as many roles an admin creates on his instance.
It would be something like, getting a list of available roles on a room, and create the specific tags for that room.
Not sure how the tag command works though.

**Comment by slyk** [comment]
+1 need this feature. 

I think admin need to create mention groups. And If he sees conflict with username, he will deal with this manually by changing user name. With good naming rules sole user wouldn't name himself using plural name ('-s' end), so conflicts will be rare.

NOT roles. Roles seems to be easy way to group people. But roles here is like system way of managing permissions. Group of users could have different chat permissions and still I want to mention them all. Connecting roles and mention groups will add unneeded restrictions.

"Users" menu is good place to configure mention groups. From there you see usernames and system can show you conflicts. It could be just sidebar button that opens sidebar with groups list. Don't think someone will have too much groups to create separate page for them.

P.S. the easy way for small teams create sort of inline alias, that will replace group name to list of usernames. Example: you wrote
"@managers"
and in chat it appeared as
"@manager1 @manager2 @manager3"

**Comment by ginnj-gilmore** [comment]
+1 above solutions seem viable and it would be a great feature to have!

**Comment by mxmo0rhuhn** [comment]
We would love this feature! 

[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=30753350)](https://www.bountysource.com/issues/30753350-notifications-implementing-custom-group-notifications?utm_source=30753350&utm_medium=shield&utm_campaign=ISSUE_BADGE)

**Comment by jgtoriginal** [comment]
After giving it a thought, it would be actually great to give users the option to tag themselves as 'developers', 'lawyers', etc. And also give admins the right to tag people from the dashboard. 
This would be only an extra attribute on the user object.
From which then we can write a hook to pick tags part of the mentions functionality.
How does that sound to you all?

**Comment by mxmo0rhuhn** [comment]
Tagging sounds like a good idea - as always LDAP integration would be awesome. 
For our organisation it would be better if the mentions would not be expanded. Imagine a team with 20 people or so would be expanded like: 
```@devs could you have a look?```
```@dev1 @dev2 @dev3 ... @dev20 could you have a look?```

**Comment by KrystalMoore16** [comment]
+1 please implement this feature!

**Comment by jgtoriginal** [comment]
after having a look at the code behind this, I think it would be highly advantageous to give `owners` and `admins` the option to turn off the `@all` feature on a channel basis. 
So say you go to `#support` and want to call everybody's attention, but your question is not as important to every single user there, `admins` would have the possibility to disallow `@all` and only allow `@here` or `@some-random-tem`

**Comment by KrystalMoore16** [comment]
Any update? Thanks!

**Comment by vynmera** [comment]
Perhaps we could use a different character, for example +admin? They can be role-based, but you can turn mentionability on and off per role. We can add an extra field to the role settings to how they should be displayed in chat, so that the "Admin" role will be displayed as +Admins.

**Comment by BanHammor** [comment]
I would really like this feature implemented, +1

**Comment by bwesemann-cxp** [comment]
+1 would love this feature. `@all` is getting annoying in large chats

**Comment by Wirone** [comment]
In our organization we have channels per team (public, so everybody can ask each team about something), sometimes per project (members are team that maintain project + invited people from business etc.). Most of the time there is some part of members that is responsible for the channel (let's call it "owners") and rest of members are guests. This feature would give us ability to call such "owners" without need to notify all channel members.

For example, there is channel for Jira/Confluence admins. Everybody can ask there something, but using `@all` or `@here` would notify too many people as we want to only draw Jira team's attention. Currently we can't easily mention group of users and mentioning each one is too cumbersome.

**Comment by FaizanZahid** [comment]
i would love to have this feature +1

**Comment by slyk** [comment]
Maybe someone will add some more coins to bounty? And then comunity and devs will understand how this important?)

Waiting it for a long time.

**Comment by gpoudrel** [comment]
Great feature it would be.

**Comment by Atomicbeast101** [comment]
Is this feature in consideration of being implemented at some point in the future?

**Comment by slyk** [comment]
Wow, so good bounty. I wish I know that programming language to write this feature by myself :)

**Comment by philtweir** [comment]
Now over £400, at least if completed in next couple months :)

**Comment by maciek134** [comment]
Is someone working on this? I'll gladly implement that feature as I need it myself.

**Comment by maciek134** [comment]
Ok, bountysource wants a link to code so I'll do that when I have something substantial to show, for now I'll just let you know here that I'm working on it :)

**Comment by feltra** [comment]
This feature has been forgotten ?

**Comment by maciek134** [comment]
Not by me, it's just going slowly as I don't have much free time atm.

**Comment by Kailash0311** [comment]
Can I also work on it? @maciek134 , I am interested in implementing the feature.

**Comment by slyk** [comment]
wow wow, we have two minds and 4 hands on this task? this is great!
this is long awaited feature, hope you can somehow cooperate with each other, to not make same work twice. Maybe split by UI and server logic?

**Comment by raatti** [comment]
+2 for this, desperately needed for any good group chat to create groups and aliases to mention.

**Comment by philtweir** [comment]
(just to highlight, at least some of that bounty is time-limited - not immediately expiring, but keen to flag)

**Comment by maciek134** [comment]
Ok, so I guess @Kailash0311 decided to post a WIP PR that adds pretty much nothing to "reserve a spot"? I'm out then, I'm not going to rush anything because of a bounty... (keep in mind that this is not a trivial feature, 90% of the work is figuring out how to do it properly).

**Comment by Kailash0311** [comment]
@maciek134 I just gave it a shot by implementing a part of this feature (there is still a lot to be left and lots of refactoring has to be done) and I am not after bounty, I did it as I think this feature is really needed :slightly_smiling_face:  
Feel free to work on your own :slightly_smiling_face: .

**Comment by slyk** [comment]
Hello. We have in mid-term milestone:
1. Group to invite #7152
2. Mention multiple users #7043
3. And current about "notifications" #2192

They almost duplicates and solving one will solve others. I'm not good expert on GitHub, but maybe all of those requests need to be closed as duplicate of current one, to redirect all requests to one discussion and have ability to make one good solution for all?

@maciek134 any news from you? Kailash told that it's not about bounty for him. Maybe you can found some time for this feature?

P.S. added a little more bounty :)
[![Bountysource](https://camo.githubusercontent.com/8566ccd998725445bdd48a7b8402a619c7fb7243/68747470733a2f2f6170692e626f756e7479736f757263652e636f6d2f62616467652f69737375653f69737375655f69643d3330373533333530)](https://www.bountysource.com/issues/30753350-notifications-implementing-custom-group-notifications?utm_source=30753350&utm_medium=shield&utm_campaign=ISSUE_BADGE)

**Comment by maciek134** [comment]
I'll take another stab at it :)

**Comment by SeaLife** [comment]
Hopefully this can be used per-channel instead of per-instance.

In a large organization with like 8000 (or nearly 20k, like in #1315) people using rocket.chat, global groups/roles to be mention will be useless. But havin a channel (like Discord has 'servers') where you can mention the @developers or @templating would help alot.

So there must be a per-channel custom notification system?

**Comment by maciek134** [comment]
I've got some free time on my hands so I'll try to have something before the end of the week.

@SeaLife yeah, that's what I was thinking - but global groups too, so you can be in a QA group but if someone mentions `@qa` in a channel you will only get notified if you already belong to it.

**Comment by SeaLife** [comment]
Did you end up with something? @maciek134 :)

**Comment by maciek134** [comment]
A lot of bad feelings for meteor ;p Seriously though, I've made some
progress - figured out the notification part, now I need to finish settings
and I'll have something to show.

**Comment by yuritoledo** [comment]
+1

**Comment by Crayder** [comment]
This would also be really awesome with an addition of the option to only notify people marked active.

Even further, you could add online-detection and these custom tags to the bot's api. 

This would help for customer notification systems for business-class users like the company I work for a lot!

**Comment by J0kerFace** [comment]
This is the mostly missing function at the moment.

**Comment by maciek134** [comment]
I have good and bad news - bad news is that our company no longer uses Rocket internally so I need to work on this in my free time only. Good news is that this is my top priority now, so you can expect this to be done shortly ;p

**Comment by maciek134** [comment]
Progress update (or more like a teaser):
![image](https://user-images.githubusercontent.com/1146309/69277537-baa24d00-0be0-11ea-9181-1b61dba44293.png)
Got right into zone after work today and finished most of the work, should have a PR ready tomorrow/friday.

**Comment by J0kerFace** [comment]
looks really cool! we are looking forward to test it out

**Comment by SeaLife** [comment]
Hopefully this can be set independentlly in each room to not have the sysadmins create notifications groups in a 8000+ employeed company :D 

But this looks really nice.

**Comment by Wirone** [comment]
@SeaLife personally I don't think notification groups should be defined per channel, or at least there should possibility to choose type of such group. We would like to make groups per team, so on every channel team could be mentioned and notified in the same way (defined once, used in multiple rooms). But on the other hand I see scenarios, where group per room could be useful - for example `maintainers` for per-project rooms.. @maciek134 what do you think?

**Comment by maciek134** [comment]
I tried to think about everything mentioned here, so I already have a solution for this (at work atm so can't post a screenshot) - when creating/editing group you have the following options:
- name
- description
- whether the group also notifies a specific role
- whether the group also notifies a specific group (so in combination with a few other options you can make a transparent copy that notifies offline users too)
- whether users can join this group themselves
- whether the group notifies offline people (like @all vs @here)
- whether the group notifies people outside a channel
- whether the group works in specific channels only
I guess there also should be an option to make it so users are also assigned per-room but I'd make it further down the line since this is a pretty specific scenario that afaik nobody asked for yet - should be relatively easy though.

**Comment by SeaLife** [comment]
I agree with this in general, but we'r using RC with over 8000 employees. Having custom groups per Channel and creatable by the channel Admin would be really nice to not bother the Admins with notifications groups. :)

**Comment by Crayder** [comment]
>whether the group notifies offline people (like @all vs @here)

Thank you!

**Comment by Wirone** [comment]
@SeaLife but with granular permissions in RocketChat it could be possible to create role for managing notification groups and allow users to do it, so Admins won't be bothered that much. I don't think notification groups configuration should go to channel's settings panel.

**Comment by SeaLife** [comment]
@Wirone okay, we should probably evaluate this first...
Its probably not that good if everyone can create global notification groups...
But we'll see :)

Im looking forward to this :)

**Comment by maciek134** [comment]
This should only take a day or two more, I re-did the admin side of things since it was a bit crammed in a side panel when editing groups.

**Comment by J0kerFace** [comment]
What's the status here?

**Comment by maciek134** [comment]
Got overwhelmed with other work, I hope to have this finished by the end of the week.

**Comment by Teqashi** [comment]
@maciek134 Any updates on this? It would help a lot

**Comment by maciek134** [comment]
@Teqashi I'm almost done, just didn't have time to wrap it up and send a PR. I'll try to do it in the evening.

**Comment by fdellwing** [comment]
We just had a meeting, discussing to supersede RC because not custom mentions are possible. Really looking forward into testing this!

**Comment by maciek134** [comment]
@fdellwing I'm pushing it today

**Comment by fdellwing** [comment]
That's correct, we looked at Slack and Teams. But we would like to further use RC, because we really like the `discussions`.

**Comment by J0kerFace** [comment]
HipChat have it too. But that's been discontinued.

**Comment by compuguy** [comment]
@maciek134 When you get a chance, can you please link to your pull request when you open it?

**Comment by maciek134** [comment]
GitHub links PRs with issue mentions automatically ;p It's pretty much done, but user self-assignment is missing along with a few translations.

**Comment by compuguy** [comment]
I didn't know that @maciek134. I understand that this is a WIP. It will be worth the wait though.

**Comment by thesystemninjaneer** [comment]
Not sure if this is the right Issue to request this but can the functionality to message using `@owner` and `@leader` of public and private channels be added as well?

We have CI/CD pipelines automatically notifying channels of pass/fail build status and the owner/leader of the channel cares about the build status the most. Support for those native roles saves us a few extra steps of having to setup a group and and/remove members over time as employees change their roles or move on in their career while the channel Leader/Owner consistently remains up to date.

**Comment by Crayder** [comment]
@thesystemninjaneer That's basically what this whole issue is suggesting, custom groups and custom notification filtering for those groups. With this implementation you'll be able to make those groups and assign people to them like normal except now you will be able to only notify active people and stuff. For being the owner of this particular group you'd just create a group for the channel in question called "owner" and assign yourself to it.

**Comment by thesystemninjaneer** [comment]
@Crayder I’m tracking the Group approach coming with this the feature in this isse but was asking for a way to notify the person assigned a particular channel role (such as owner or leader) without the extra steps of creating and maintaining a custom group. Is that something easily achievable?

**Comment by maciek134** [comment]
@thesystemninjaneer easily achievable - yes, but I don't think this is in the scope of this issue / PR.

**Comment by Shaverdoff** [comment]
any progress of that? or this develop-issues need some donate?

**Comment by maciek134** [comment]
@Shaverdoff I may finally have some time to finish this one, I'll try to push fixes this week.

**Comment by sviat9440** [comment]
Any news ?

**Comment by julianeisel** [comment]
Let me second this! Such a feature would be very helpful for as at Blender ([blender.chat](https://blender.chat/)), and allow us to move more conversations to public channels.

There's lots of activity in our public channels, if I want to notify a specific team (say the core UI-team) it's a hassle to do in public channels. Hence some teams end up using lower-traffic, private team channels more, which is against our own transparency-principles. 
We use read-only channels in some instances, but they further fragment public communication which is an issue already.

**Comment by sviat9440** [comment]
@maciek134 are you have problem? 2021 still waiting (

**Comment by vinogradovia** [comment]
Any news ?

**Comment by fdellwing** [comment]
https://github.com/RocketChat/Rocket.Chat/pull/16311#issuecomment-825389926 ff

**Comment by vinogradovia** [comment]
thx,  strange that I did not find this functionality in the price (licence comparison)

**Comment by fdellwing** [comment]
https://docs.rocket.chat/guides/user-guides/rooms/teams/teams-enterprise-edition

It's not even included in the middle price option, only in the highest price that they do not even communicate openly.
